openssl/src/pkcs7.rs

```diff
@@ -281,11 +296,47 @@ impl Pkcs7Ref {
             Ok(stack)
         }
     }
+
+    // Return the type of a PKCS#7 structure as a NID
+    pub fn type_nid(&self) -> Option<Nid> {
+        unsafe {
+            let ptr = (*self.as_ptr()).type_;
+            Asn1ObjectRef::from_const_ptr_opt(ptr).map(|x| x.nid())
+        }
+    }
+
+    // Retrieve all the certificates from a PKCS#7 structure used for signed data
+    pub fn signed_data_certificates(&self) -> Result<Stack<X509>, ErrorStack> {
+        unsafe {
+            let mut new_stack = Stack::<X509>::new()?;
+            if self.type_nid() != Some(Nid::PKCS7_SIGNED) {
+                return Ok(new_stack);
+            }
+            // Get the stack of certificates from the PKCS7_SIGNED object
+            let stack_certs = (*self.as_ptr())
+                .d
+                .sign
+                .as_mut()
+                .and_then(|x| x.cert.as_mut())
+                .and_then(|x| StackRef::<X509>::from_const_ptr_opt(x));
+            match stack_certs {
+                None => Ok(new_stack),
```

Comment:
I think you should maek this return `Result<Option<T>>` so callers can know if `.d.sign` was null (which the RFC says you're supposed to handle in a specific way)

Reply:
Changed to return a `Result<Option<&StackRef<X509>>, ErrorStack>`, so it matches better the actual type of the `.d.sign` member (a pointer to a stack)